Show both ClusterResourceQuotas and ResourceQuotas

rebeccaalpert opened the pull request:
Fixes [CONSOLE-653](https://jira.coreos.com/browse/CONSOLE-653).

Modified ResourceQuotaPage to show both ClusterResourceQuotas and
ResourceQuotas.

<img width="1238" alt="screen shot 2018-08-24 at 4 36 33 pm" src="https://user-images.githubusercontent.com/7014965/44606680-e9eac400-a7bb-11e8-8e5a-818e9d375224.png">

rebeccaalpert commented:
jenkins rebuild

spadgett commented on frontend/public/components/resource-quota.jsx:
```diff
@@ -1,38 +1,30 @@
+  const kind = quota.metadata.namespace ? 'ResourceQuota' : 'quota.openshift.io:v1:ClusterResourceQuota';
+  return <React.Fragment>
+    <ResourceCog actions={menuActions} kind={kind} resource={quota} />
+    <ResourceLink kind={kind} name={quota.metadata.name} namespace={quota.metadata.namespace} className="co-resource-link__resource-name" />
```
I believe you can make this `kind={quota.kind}` if you add a model to

https://github.com/openshift/console/blob/master/frontend/public/models/index.ts

Then you wouldn't need the special component for name.

spadgett commented on frontend/public/components/resource-quota.jsx:
```diff
@@ -1,38 +1,30 @@
 
 const menuActions = [Cog.factory.ModifyLabels, Cog.factory.ModifyAnnotations, Cog.factory.Edit, Cog.factory.Delete];
 
+export const QuotaName = ({quota: quota}) => {
```
You shouldn't need to `export` if this is only used in the same file

spadgett commented on frontend/public/components/resource-quota.jsx:
```diff
@@ -1,38 +1,30 @@
 const menuActions = [Cog.factory.ModifyLabels, Cog.factory.ModifyAnnotations, Cog.factory.Edit, Cog.factory.Delete];
 
+export const QuotaName = ({quota: quota}) => {
+  const kind = quota.metadata.namespace ? 'ResourceQuota' : 'quota.openshift.io:v1:ClusterResourceQuota';
```
You should be able to use `referenceFor` (but it also shouldn't be needed if you register the ClusterResourceQuota model)

https://github.com/openshift/console/blob/2c7a4b0b31a9046ad03863382a968e386273b6db/frontend/public/module/k8s/k8s.ts#L18

rebeccaalpert commented:
I need to use the namespace to determine the kind, but I added the model, got rid of the special component for name, and plugged in referenceFor. Let me know if adding the model is unnecessary.

spadgett commented on frontend/public/components/resource-quota.jsx:
```diff
@@ -29,7 +33,54 @@ const Details = ({obj: rq}) => <React.Fragment>
+    rows.push(quota);
+  });
+
+  return rows;
```
Assuming resource.data is an array, I think this can be... (untested)

```js
export const flatten = resources => _.flatMap(resources, resource => _.compact(resource.data));
```

spadgett commented on frontend/public/components/resource-quota.jsx:
```diff
@@ -29,7 +33,54 @@ const Details = ({obj: rq}) => <React.Fragment>
+  return <MultiListPage
+    canCreate={true}
+    createButtonText="Create Resource Quota"
+    createProps={{to: '/k8s/ns/tectonic-system/resourcequotas/new'}}
```
Why is this using `tectonic-system`?

rebeccaalpert commented:
I added flag checking and fixed up the function/link.

spadgett commented on frontend/public/components/resource-quota.jsx:
```diff
@@ -29,7 +35,56 @@ const Details = ({obj: rq}) => <React.Fragment>
+    createButtonText="Create Resource Quota"
+    createProps={{to: `/k8s/ns/${ns}/resourcequotas/new`}}
+    ListComponent={ResourceQuotasList}
+    staticFilters={[{'resource-quota-roleRef': name}]}
```
I think this was incorrectly copied from the role binding page.

spadgett commented on frontend/public/components/resource-quota.jsx:
```diff
@@ -29,7 +35,56 @@ const Details = ({obj: rq}) => <React.Fragment>
+    namespace={ns}
+    flatten={flatten}
+    title="Resource Quotas"
+    rowFilters={[{
```
Row filters should only be supplied if this is OpenShift. Otherwise there's only one kind of quota and no need for the filter.

spadgett commented:
Looks like the integration tests need to be updated for this change.

rebeccaalpert commented:
Removed static filters and made it so row filters are only supplied if it's OpenShift. The integration test was failing because I'd broken the search page for Resource Quotas. I fixed it and the crud test is running OK locally.

spadgett commented:
jenkins rebuild

spadgett commented:
Here's the performance test error:

```
1) Performance test : downloads new bundle for custom-resource-definition
   TypeError: Cannot read property 'decodedBodySize' of null
       at Object.<anonymous> /go/src/github.com/openshift/console/frontend/integration-tests/tests/performance.scenario.ts:91:25
       at step /go/src/github.com/openshift/console/frontend/integration-tests/tests/performance.scenario.ts:32:23
       at Object.next /go/src/github.com/openshift/console/frontend/integration-tests/tests/performance.scenario.ts:13:53
       at fulfilled /go/src/github.com/openshift/console/frontend/integration-tests/tests/performance.scenario.ts:4:58
       at process._tickCallback internal/process/next_tick.js:68:7
   From: Task: Run it"downloads new bundle for custom-resource-definition" in control flow
       at UserContext.<anonymous> /go/src/github.com/openshift/console/frontend/node_modules/jasminewd2/index.js:94:19
   From asynchronous test: 
   Error
       at /go/src/github.com/openshift/console/frontend/integration-tests/tests/performance.scenario.ts:84:5
       at /go/src/github.com/openshift/console/frontend/node_modules/immutable/dist/immutable.js:2699:43
       at List.__iterate /go/src/github.com/openshift/console/frontend/node_modules/immutable/dist/immutable.js:2206:13
       at OrderedMap.__iterate /go/src/github.com/openshift/console/frontend/node_modules/immutable/dist/immutable.js:2698:25
       at OrderedMap.forEach /go/src/github.com/openshift/console/frontend/node_modules/immutable/dist/immutable.js:4381:19
       at Suite.<anonymous> /go/src/github.com/openshift/console/frontend/integration-tests/tests/performance.scenario.ts:76:17
       at Env.<anonymous> /go/src/github.com/openshift/console/frontend/node_modules/jasmine-fail-fast/dist/jasmine-fail-fast.js:57:26
```

spadgett commented:
I can't explain why you're seeing that error with these changes :(

cc @alecmerdler

alecmerdler commented:
That failure indicates that a separate chunk for the "custom-resource-definition" route isn't being downloaded (or has a different name).

spadgett commented:
jenkins rebuild
